### PR TITLE
PR83 reinforcement learning with alpha zero

### PR DESCRIPTION
**Pull request by NaniPuerta**
## Integrate AlphaZero reinforcement learning algorithm to autogoal

Reinforcement learning is a broad claim to add a PR. Here we feature an implementation of the AlphaZero algorithm able to auto train and generate a virtual player for any game that requires 2 players and can be played in a game board, such as Chess, Hex, Connect4, TicTacToe, Go and others.

In this project we intended to create a general AlphaZeroX, such that given a "Game" we generate an agent capable of playing that game and integrate that algorithm into AutoGoal semantics type system, in order to be able to select that algorithm to be part of the used pipelines.

That being said, we provide a Game class, which serves as an interface for any Game definition that match our algorithm input. This is an abstract class that enforce some restrictions over the game. The most important one here, is that of symetry, i.e, the property of a Game to be equally good for player A than for player B if player A and B swap positions. This means that in a tic tac toe game, is equally good for one player, if she is in a winning position, no matter if she is running X or O. The same goes for chess if players swap positions or exchange colors of pieces. Not all boards are trivially reversed, but that's left to the Interface Implementators.

We provide an AlphaZeroAgent, which basically takes a Game and a Neural Network and spits out a player. The game this agent requires is one that meet the Game interface. In this class the main method is train. This method uses the given NN, which is a Convolutional Neural Network, to perform searches over a Monte Carlo Tree Search, and training over the collected examples of each run, basically, trying to predict the win probability of a play among all posibles from a given board state. Then the NN, plays against the best performant NN that has been trained so far, and we keep the best.

AlphaZero is beatiful, but it depends on a bunch of parameters that are optimally different for every game, and that is when AutoGoal comes in handy.

In order to integrate the algorithm with autogoal, we defined 3 Semantic Types:

- Environment: A game could be seen as a multidimensional environment, where resources are located, and there are predefined actions over the environment that produces a reward. The whole idea of a game, is to select the best possible sequence of actions to obtain the higher reward. That being said, our implementationt only deals with 2-D environments, and those must conform to the Game Interface. So we define an Environment Semantic type for future generalization of our algorithm.
- GameStructure: Semantic type that match only subclasses of Game interface
- Player: return type that match classes with a play method that receive a game and makes an action over that game.

Given this semantics, we defined the AlphaZeroAlgorithm class, inheriting from AlgorithmBase, where in the constructor, we suplied hyperparameters for the CNN construction and training, such as learning rate, memory, epochs, arena_games (matches), etc. This class overrides the run method to annotate the inputs and outputs of our algorithm and futhermore, register receive methods for train and eval modes.

Games are not easily scored, our algorithm does not return a float value we could compare with, instead we receive a player. For metrics purpose, we defined a meta-metric function (a function that returns a metric), that basically receives a game, and returns the win score of the resulting player against a predefined player. Here, win score is a function that receives two players, and pits them in an arena of 100 games and returns the win/games values for the trained player.

With this, generating a player is a matter of time and defining the game we want. As examples, we provide implementations for Tic Tac Toe and Connect4 games.

Using this with autogoal, is as easy as write:

```python
game = TicTacToeGame()
pred_player = RandomPlayer(game)
metric = create_win_score_metric(game)
ml = AutoML(
       input = GameStructure,
       output = Player,
       registry = [AlphaZeroAlgorithm]
       score_metric = metric
)

ml.fit(game)

score = ml.score(game, pred_player)
print(score)
```
As a full example, we provide this script, which train a player on TicTacToe and scored her against a Random Player.
In this example, we are overriding AlphaZeroAlgorithm to tweek the parameters a little bit, so the player does not take, lets say, a day, to train. Nontheless, for good performance of the player, it is advised that the algorithm runs for at least a day, depending on the complexity of the game.
```python
from autogoal.ml import AutoML
from autogoal.experimental.alphazero import create_win_rate_metric
from .TicTacToeGame import TicTacToeGame
from .ramdomAgent import RandomPlayer
from autogoal.experimental.alphazero import GameStructure, Player
from autogoal.experimental.alphazero import AlphaZeroAlgorithm
from autogoal.grammar import ContinuousValue, DiscreteValue, CategoricalValue


def run_example():

    game = TicTacToeGame()

    test_player = RandomPlayer(game)

    win_rate = create_win_rate_metric(game, display=game.display)

    # Standard RL training is a long process, trying to improve with
    # the maximun simulations posible. For a simple game like Tic Tac Toe
    # we might use a very simple training process, and reduce training time to just
    # 5 minutes Top

    class AlphaZeroTicTacToe(AlphaZeroAlgorithm):
        def __init__(
            self,
            reg_const: ContinuousValue(0.0001, 0.001),
            learning_rate: ContinuousValue(0.0001, 0.02),
            batch_size: CategoricalValue(8, 16, 32, 64, 128, 256),
            epochs: DiscreteValue(10, 100),
            num_iters: DiscreteValue(10, 15),
            queue_len: DiscreteValue(20000, 40000),
            episodes: DiscreteValue(2, 5),
            memory_size: DiscreteValue(3000, 10000),
            arena_games: DiscreteValue(40, 100),
            update_threshold: ContinuousValue(0.5, 0.9),
            time_limit: DiscreteValue(5 * 60, 10 * 60),  # limit to 5-10 minutes
        ) -> None:
            super().__init__(
                reg_const,
                learning_rate,
                batch_size,
                epochs,
                num_iters,
                queue_len,
                episodes,
                memory_size,
                arena_games,
                update_threshold,
                time_limit,
            )

    automl = AutoML(
        input=GameStructure,
        output=Player,
        score_metric=win_rate,
        registry=[AlphaZeroTicTacToe],
    )

    automl.fit(X=game, y=test_player)

    score = automl.score(game, test_player)

    print(score)
```

# New Files
- autogoal/experimental/alphazero/MTCS.py : Monte Carlo Tree Search Implementation
- autogoal/experimental/alphazero/Agent.py: AlphaZeroAgent implementation
- autogoal/experimental/alphazero/GameInterface.py
- autogoal/experimental/alphazero/GameManager.py : Facilities to play matches
- autogoal/experimental/alphazero/NeuralNet.py: CNN implementation
- autogoal/experimental/_semantics.py: New semantics types for Algorithm integration
- autogoal/experimental/_metrics.py: Meta metric function generator
- autogoal/experimental/_algorithm.py: Register algorithm in AutoGoal

# Semantic Types:
- Enviroment: base abstraction for games.
- GameStructure: Recognizes a 2-D game that conforms to Game interface.
- Player: Return type of algorithm. Match classes with a play method.

# Examples
- autogoal/tests/experimental/alphazero/Connect4Game.py: Connect4 implementation using Game interface.
- autogoal/tests/experimental/alphazero/randomAgent.py: Random player for any game implementation.
- autogoal/tests/experimental/alphazero/TicTacToeGame.py: Tic Tac Toe game implementation using Game interface.
- autogoal/tests/experimental/alphazero/test_alphaZeroAlgorithm_example.py: Runs an example of AutoML using our algorithm to train a Tic Tac Toe player.

### CODE REVIEW

**Comment by adriangs1996** [comment]
Reinforcement learning is a broad claim to add a PR. Here we feature an implementation of the AlphaZero algorithm able to auto train and generate a virtual player for any game tha requires 2 players and can be played in a game board, such as Chess, Hex, Connect4, TicTacToe, Go and others.

In this project we intended to create a general AlphaZeroX, such that given a "Game" we generate an agent capable of playing that game and integrate that algorithm into AutoGoal semantics type system, in order to be able to select that algorithm to be part of the used pipelines.

That being said, we provide a Game class, which serve as an interface for any Game definition that match our algorithm input. This is an abstract class that enforce some restrictions over the game. The most important one here, is that of symetry, i.e, the property of a Game to be equally good for player A than for player B if player A and B swap positions. This means that in a tic tac toe game, is equally good for one player, if she is in a winning position, no matter if she is running X or O. The same goes for chess if players swap positions or exchange colors of pieces. Not all boards are trivially reversed, but that's left to the Interface Implementators.

We provide an AlphaZeroAgent, which basically takes a Game and a Neural Network and spits out a player. The game this agent requires is one that meet the Game interface. In this class the msin method is train. This method uses the given NN, which is a Convolutional Neural Network, to perform searches over a Monte Carlo Tree Search, and training over the collected examples of each run, basically, trying to predict the win probability of a play among all posibles from a given board state. Then the NN, plays against the best performant NN that has been trained so far, and we keep the best.

AlphaZero is beatifull, but it depends on a bunch of parameters that are optimally different for every game, and that is when AutoGoal comes in handy. 

In order to integrate the algorithm with autogoal, we defined 3 Semantic Types:

- Environment: A game could be seen as a multidimensional environment, where resource are located, and there are predefined actions over the environment that produces a reward. The whole idea of a game, is to select the best possible sequence of actions to obtain the higher reward. That being said, our implementationt only deals with 2-D environments, and those must conform to the Game Interface. So we define an Environment Semantic type for future generalization of our algorithm.
- GameStructure: Semantic type that match only subclasses of Game interface
- Player: return type that match classes with a play method that receive a game and makes an action over that game.
Given this semantics, we defined the AlphaZeroAlgorithm class, inheriting from AlgorithmBase, where in the constructor, we suplied hyperparameters for the CNN construction and training, such as learning rate, memory,  epochs, arena_games (matches), etc. This class overrides the run method to annotate the inputs and outputs of our algorithm and futhermore, register receive methods for train and eval modes.

Games are not easily scored, our algorithm does not return an float value we could compare with, instead we receive a player. For metrics purpose, we defined a meta-metric function (a function that returns a metric), that basically receives a game, and returns the win score of the resulting player against a predefined player. Here, win score is a function that recives two players, and pit them in an arena of 100 games and return the win/games values for the trained player.

With this, generste a player is a matter of time and define the game we want. As example, we peovide implementations for Tic Tac Toe and Connect4 games.

Using this with autogoal, is as easy as write:

```python
game = TicTacToeGame()
pred_player = RandomPlayer(game)
metric = create_win_score_metric(game)
ml = AutoML(
       input = GameStructure,
       output = Player,
       registry = [AlphaZeroAlgorithm]
       score_metric = metric
)

ml.fit(game)

score = ml.score(game, pred_player)
print(score)
```

**Comment by adriangs1996** [comment]
# New Files

- autogoal/experimental/alphazero/MTCS.py : Monte Carlo Tree Search Implementation
- autogoal/experimental/alphazero/Agent.py: AlphaZeroAgent implementation
- autogoal/experimental/alphazero/GameInterface.py
- autogoal/experimental/alphazero/GameManager.py : Facilities tu play matches
- autogoal/experimental/alphazero/NeuralNet.py: CNN implementation
- autogoal/experimental/_semantics.py: New semantics types for Algorithm integration
- autogoal/experimental/_metrics.py: Meta metric function generator
- autogoal/experimental/_algorithm.py: Register algorithm in AutoGoal

# Semantic Types:
- Enviroment: base abstraction for games.
- GameStructure: Recognizes a 2-D game that conforms to Game interface.
- Player: Return type of algorithm. Match classes with a play method.

# Examples
- autogoal/tests/experimental/alphazero/Connect4Game.py: Connect4 implementation using Game interface.
- autogoal/tests/experimental/alphazero/randomAgent.py: Random player for any game implementation.
- autogoal/tests/experimental/alphazero/TicTacToeGame.py: Tic Tac Toe game implementation using Game interface.
- autogoal/tests/experimental/alphazero/test_alphaZeroAlgorithm_example.py: Runs an example of AutoML using our algorithm to train a Tic Tac Toe player.

**Comment by apiad** [comment]
Revisado. Pásenme nombre completo de los miembros del equipo.
